test/integration/dnsrecord/dnsrecord_test.go

```diff
@@ -0,0 +1,294 @@
+
+package dnsrecord_test
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/service/route53"
+	awsinstall "github.com/gardener/gardener-extension-provider-aws/pkg/apis/aws/install"
+	"github.com/gardener/gardener-extension-provider-aws/pkg/aws"
+	awsclient "github.com/gardener/gardener-extension-provider-aws/pkg/aws/client"
+	dnsrecordctrl "github.com/gardener/gardener-extension-provider-aws/pkg/controller/dnsrecord"
+
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	"github.com/gardener/gardener/pkg/extensions"
+	gardenerutils "github.com/gardener/gardener/pkg/utils"
+	"github.com/gardener/gardener/test/framework"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	. "github.com/onsi/gomega/gstruct"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/envtest"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+var (
+	accessKeyID     = flag.String("access-key-id", "", "AWS access key id")
+	secretAccessKey = flag.String("secret-access-key", "", "AWS secret access key")
+	zoneName        = flag.String("zone-name", "", "AWS Route53 zone name")
+	zoneID          = flag.String("zone-id", "", "AWS Route53 zone ID")
+)
+
+func validateFlags() {
+	if len(*accessKeyID) == 0 {
+		panic("need an AWS access key id")
+	}
+	if len(*secretAccessKey) == 0 {
+		panic("need an AWS secret access key")
+	}
+	if len(*zoneName) == 0 {
+		panic("need an AWS Route53 zone name")
+	}
+	if len(*zoneID) == 0 {
+		panic("need an AWS Route53 zone ID")
+	}
+}
+
+var _ = Describe("DNSRecord tests", func() {
+	var (
+		ctx = context.Background()
+
+		logger    *logrus.Entry
+		awsClient *awsclient.Client
+		testEnv   *envtest.Environment
+		mgrCancel context.CancelFunc
+		c         client.Client
+
+		testName  = fmt.Sprintf("aws-dnsrecord-it--%s", randomString())
+		namespace = &corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: testName,
+			},
+		}
+		secret = &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "dnsrecord",
+				Namespace: testName,
+			},
+			Data: map[string][]byte{
+				aws.AccessKeyID:     []byte(*accessKeyID),
+				aws.SecretAccessKey: []byte(*secretAccessKey),
+			},
+		}
+		cluster = &extensionsv1alpha1.Cluster{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: testName,
+			},
+		}
+	)
+
+	BeforeSuite(func() {
+		repoRoot := filepath.Join("..", "..", "..")
+
+		// enable manager logs
+		logf.SetLogger(zap.New(zap.UseDevMode(true), zap.WriteTo(GinkgoWriter)))
+
+		log := logrus.New()
+		log.SetOutput(GinkgoWriter)
+		logger = logrus.NewEntry(log)
+
+		By("starting test environment")
+		testEnv = &envtest.Environment{
+			UseExistingCluster: pointer.BoolPtr(true),
```

Comment:
This test could also run against a temp envtest cluster and we don't necessarily need an existing cluster, right?

Reply:
Not sure, will try it out. But what's the problem to use an existing cluster? This can be really any cluster, including our Nodeless dev setup, and in CI such a cluster is already created anyway for the infra tests. The bastion tests use an existing cluster as well.

Reply:
There is no problem with it. But generally it would be possible. One can configure it via the env var `USE_EXISTING_CLUSTER`.
We could leave it off for the local case, so that one doesn't have to think about starting a cluster / pointing `KUBECONFIG` to it. In CI / TM we could specify `USE_EXISTING_CLUSTER`, if we see a benefit in it.

Just explaining the options... Feel free to choose what makes sense from your point of view.

Reply:
I tried to set the above to `false`, but I got:
```
unable to start the controlplane	{"tries": 0, "error": "fork/exec /usr/local/kubebuilder/bin/etcd: no such file or directory"}
```
I guess I need to install `kubebuilder`, but then it should also be installed via `install requirements`, and this would introduce an important difference with `install requirements` for other provider extensions. And I remember from previous experience with SEN that installing / updating kubebuilder automatically wasn't a great experience.
Are you sure we really need this, or perhaps I am missing something?

Reply:
In any case, you are right that this line should be removed, since there is the `USE_EXISTING_CLUSTER` variable that could be just set to true (and I can confirm this works as expected), I am just not really sure if I wanted to to use an existing cluster locally what are my options - just install kubebuilder locally (but not via `make install-requirements`)? Is this documented somewhere?

Reply:
Removed the line as suggested.

Reply:
We don't need to install kubebuilder, `setup-envtest` is enough, see [Makefile](https://github.com/gardener/gardener/blob/3806079d4e35dca8064e39ce0e69a0912aa02778/Makefile#L177) and [`tools.go`](https://github.com/gardener/gardener/blob/3806079d4e35dca8064e39ce0e69a0912aa02778/hack/tools.go#L28) in g/g.
`setup-envtest` can then be used to pull envtest binaries (e.g. etcd). It's already done in [`test-integration.sh`](https://github.com/gardener/gardener/blob/master/hack/test-integration.sh), so you can basically replace the `go test` call in the Makefile here with the vendored hack script and then everything should directly work.

Reply:
So basically something like this: https://github.com/gardener/gardener-extension-provider-aws/commit/0601dd8915c252efb9ad80bb322295195c7f171f

Reply:
Ah, got it, thanks!

Reply:
Added the changes you suggested in a separate commit to this PR. I can confirm everything works correctly now both with and without `USE_EXISTING_CLUSTER=true`.
/squash